README.md
# node-artedi: client library for metric collection

## About
`artedi` is a Node.js library for measuring applications -- specifically, the
services composing Triton and Manta.

## Sample Usage
Here is a simple example usage of counters and histograms to expose
metrics in the Prometheus v0.0.4 text format.

```javascript
var artedi = require('artedi');

// collectors are the 'parent' collector.
var collector = artedi.createCollector();

// counters are a 'child' collector.
// This call is idempotent.
var counter = collector.counter({
    name: 'http_requests_completed',
    help: 'count of muskie http requests completed',
    labels: {
        zone: 'e5d3'
    }
});

// Add 1 to the counter with the labels 'method=getobject,code=200'.
counter.increment({
    method: 'getobject',
    code: '200'
});

collector.collect(artedi.FMT_PROM, function (err, metrics) {
    console.log(metrics);
    // Prints:
    // # HELP http_requests_completed count of muskie http requests completed
    // # TYPE http_requests_completed counter
    // http_requests_completed{zone="e5d3",method="getobject",code="200"} 1
});

var histogram = collector.histogram({
    name: 'http_request_latency_seconds',
    help: 'latency of muskie http requests',
    buckets: [0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10]
});

// Observe a latency of 998ms for a 'putobjectdir' request.
histogram.observe(0.998, {
    method: 'putobjectdir'
});

// For each bucket, we get a count of the number of requests that fall
// below or at the latency upper-bound of the bucket.
// This output is defined by Prometheus.
collector.collect(artedi.FMT_PROM, function (err, metrics) {
    if (err) {
        throw new Error('could not collect metrics');
    }
    console.log(metrics);
    // Prints:
    // # HELP http_requests_completed count of muskie http requests completed
    // # TYPE http_requests_completed counter
    // http_requests_completed{method="getobject",code="200",zone="e5d3"} 1
    // # HELP http_request_latency_seconds latency of muskie http requests
    // # TYPE http_request_latency_seconds histogram
    // http_request_latency_seconds{method="putobjectdir",le="0.005"} 0
    // http_request_latency_seconds{method="putobjectdir",le="0.01"} 0
    // http_request_latency_seconds{method="putobjectdir",le="0.025"} 0
    // http_request_latency_seconds{method="putobjectdir",le="0.05"} 0
    // http_request_latency_seconds{method="putobjectdir",le="0.01"} 0
    // http_request_latency_seconds{method="putobjectdir",le="0.25"} 0
    // http_request_latency_seconds{method="putobjectdir",le="0.5"} 0
    // http_request_latency_seconds{method="putobjectdir",le="1"} 1
    // http_request_latency_seconds{method="putobjectdir",le="2.5"} 1
    // http_request_latency_seconds{method="putobjectdir",le="5"} 1
    // http_request_latency_seconds{method="putobjectdir",le="10"} 1
    // http_request_latency_seconds{le="+Inf",method="putobjectdir"} 1
    // http_request_latency_seconds_count{method="putobjectdir"} 1
    // http_request_latency_seconds_sum{method="putobjectdir"} 0.998
});
```

For more advanced usage and full API documentation, see
[docs/API.md](./docs/API.md).

## Installation
```
npm install artedi
```

## DTrace probes
artedi includes some useful DTrace probes. The full listing of probes and their
arguments can be found in the [lib/provider.js](./lib/provider.js) file.

In this first example artedi is observing the latency of queries to three
Postgres instances (using [TritonDataCenter/pgstatsmon](https://github.com/TritonDataCenter/pgstatsmon)).
The latency observations include the name of the backend Postgres instance.

We can create a graph of each Postgres backend's latency using built-in DTrace
aggregation:
```
$ dtrace -qn 'artedi*:::histogram-observe {@lat[json(copyinstr(arg2), "name")] = quantize(arg1);} tick-10s {printa(@lat);}'

  3.postgres.walleye.kkantor.com-87eb177c
           value  ------------- Distribution ------------- count
               8 |                                         0
              16 |@@@@@@@@@@                               5
              32 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@            15
              64 |@@                                       1
             128 |                                         0

  2.postgres.walleye.kkantor.com-335c1a83
           value  ------------- Distribution ------------- count
               8 |                                         0
              16 |@@@@@@@@                                 4
              32 |@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@           16
              64 |@@                                       1
             128 |                                         0

  1.postgres.walleye.kkantor.com-f5c49b33
           value  ------------- Distribution ------------- count
               8 |                                         0
              16 |@@@@@@@@                                 4
              32 |@@@@@@@@@@@@@@@@@@@@@@@@@@@              14
              64 |@@@@@@                                   3
             128 |                                         0
^C
```

We could also retrieve the number of HTTP operations performed by HTTP handler
name and return code (from manta-muskie):
```
$ dtrace -qn 'artedi*:::counter-add /copyinstr(arg0) == "http_requests_completed" /{ jsonstr = copyinstr(arg2); @counts[json(jsonstr, "operation"), json(jsonstr, "statusCode")] = count(); }'
^C

  getstorage    200     135
  putobject     204     137
```

These probes could conceivably be used to create more complicated reporting
tools as well.

## License

MPL-v2


CHANGES.md
# Changelog

## Not yet released

None

## 2.0.4

- TRITON-2325 CVE-2021-3918: json-schema is vulnerable to Prototype Pollution

## 2.0.3

- TRITON-2254 Change joyMattermostNotification to joySlackNotification

## 2.0.2

- MANTA-5184 bump dtrace-provider dep and relax other deps

## 2.0.1

* MANTA-4959 Use Math.log10() ployfill

## 2.0.0 [backward incompatible if you use histograms]

* #17 Histogram buckets completely changed to be compatible with Prometheus. You
  can now either pass in an array of bucket values. If you do not pass in bucket
  values, you will get the default buckets:
    ```
    [0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5, 10]
    ```
* Important: see [docs/migrating.md](docs/migrating.md) for details on migrating
  from node-artedi v1.x to v2.x.

## 1.4.1
* #15 improve the performance of hashObj()

## 1.4.0
* #14 Allow for metrics in a MetricVector to be optionally expired or reset to a
    default value

## 1.3.0
* #13 add DTrace probes

## 1.2.1
* #12 histograms can't observe values less than one

## 1.2.0
* #9 Create an accessor function for collector values

## 1.1.1
* [MORAY-447](https://smartos.org/bugview/MORAY-447) strange latency data point
    in reported moray metrics

## 1.1.0
* #5 Implement set() for Gauge collectors
* #6 Implement basic triggered metrics
* #7 artedi shouldn't attach timestamps to prometheus metrics
* #8 artedi collectors should share more code

## 1.0.0
* #3 Support Prometheus-style counters and histograms


lib/histogram.js
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * Copyright (c) 2018, Joyent, Inc.
 */
var mod_util = require('util');

var mod_assert = require('assert-plus');
var mod_jsprim = require('jsprim');
var VError = require('verror').VError;

var lib_utils = require('./utils');
var lib_counter = require('./counter');
var lib_gauge = require('./gauge');
var lib_common = require('./common');
var lib_provider = require('./provider');


/*
 * These default buckets match the official golang, javascript, rust and other
 * client libraries.
 */
var DEFAULT_BUCKETS = [
    0.005,
    0.01,
    0.025,
    0.05,
    0.1,
    0.25,
    0.5,
    1,
    2.5,
    5,
    10
];


/*
 * A Histogram is a type of collector that represents a series of Counters. Each
 * Counter corresponds to a certain range of values, called 'buckets.'
 */
function Histogram(options) {
    mod_assert.object(options, 'options');
    mod_assert.string(options.name, 'options.name');
    mod_assert.string(options.help, 'options.help');
    mod_assert.optionalArrayOfNumber(options.buckets, 'options.buckets');
    mod_assert.optionalObject(options.labels, 'options.labels');
    mod_assert.optionalObject(options.parentLabels, 'options.parentLabels');

    var i;
    var prevBucket = -1;

    this.staticLabels =
        mod_jsprim.mergeObjects(options.parentLabels, options.labels, null);

    this.name = options.name;
    this.help = options.help;
    this.type = lib_common.HISTOGRAM;

    this.buckets = options.buckets || DEFAULT_BUCKETS;
    this.counters = {};
    this.gauge = new lib_gauge.Gauge(options);

    // Assert that buckets are monotonic
    for (i = 0; i < this.buckets.length; i++) {
        mod_assert.ok(this.buckets[i] > prevBucket,
            'buckets should be monotonic [' + this.buckets[i] +
            ' > ' + prevBucket + ']');
        prevBucket = this.buckets[i];
    }
}

/* Public Functions */
/*
 * Determine which bucket the observed 'value' falls into, and increment all the
 * Counters >= that observed value. The '+Inf' Counter is always incremented,
 * and a Gauge is created to track the running sum of values observed.
 */
Histogram.prototype.observe = function observe(value, pairs) {
    mod_assert.number(value, 'value');
    mod_assert.optionalObject(pairs, 'pairs');

    var name = this.name;
    lib_provider['histogram-observe'].fire(function () {
        return ([name, value, pairs]);
    });

    if (value < 0) {
        throw new VError('observe must be called with a value >= 0: %d', value);
    }

    var i;
    var counter;
    var pairCopy;

    pairs = mod_jsprim.mergeObjects(pairs, this.staticLabels, null);
    counter = this.labels(pairs);

    // Make a copy of the labels sent in.
    pairCopy = mod_jsprim.deepCopy(pairs);

    // Increment the counters for each bucket(le) where "value" is <= the bucket
    for (i = 0; i < this.buckets.length; i++) {
        pairCopy['le'] = this.buckets[i];
        if (value <= this.buckets[i]) {
            counter.increment(pairCopy);
        } else {
            // We don't need to increment buckets that are smaller than value,
            // but we do want to zero them out if they don't exist.
            counter.labels(pairCopy);
        }
    }

    // Always increment the Inf Counter.
    counter.increment({
        le: '+Inf'
    });

    // There is a gauge for each counter so we can keep track
    // of the _sum field (which can move up or down).
    this.gauge.add(value, counter.staticLabels);
};

/* Private Functions */
/*
 * Returns a Counter associated with the givel labels, 'pairs,' and initializes
 * a Gauge which will be used to track the sum of the values added to the
 * Counter.
 */
Histogram.prototype.labels = function labels(pairs) {
    mod_assert.optionalObject(pairs);
    if (!pairs || mod_jsprim.isEmpty(pairs)) {
        return (this.defaultCounter());
    }
    var lhash;
    var opts;
    var my_counter;

    pairs = lib_utils.trim(pairs);
    lhash = lib_utils.hashObj(pairs);
    if (this.counters[lhash]) {
        // We have already recorded this Counter.
        return (this.counters[lhash]);
    }

    // Before we create the Counter, append the static labels.
    pairs = mod_jsprim.mergeObjects(pairs, this.staticLabels, null);

    // Create a new Counter.
    opts = {
        name: this.name,
        help: this.help,
        labels: pairs
    };
    my_counter = new lib_counter.Counter(opts);
    this.gauge.labels(my_counter.staticLabels); // Initialize the gauge.

    this.counters[lhash] = my_counter;
    return (my_counter);
};

/*
 * The 'prometheus()' function for Histograms is more complicated than that of
 * Counters and Gauges.
 *
 * Each Counter represents a number of Metrics (via MetricVectors), so we
 * call 'prometheus()' on each Counter's underlying MetricVector object. After
 * we serialize each Counter, we synthesize the '_count' field and append the
 * '_sum' field. The '_count' field carries the same value as the corresponding
 * Counter's '+Inf' label.
 */
Histogram.prototype.prometheus = function prometheus(cb) {
    mod_assert.func(cb, 'callback');
    var str = '';
    var labelStr = '';
    var counter;
    var keys;
    var countLabels, key, labelObj, label;

    var infHash = lib_utils.hashObj({le: '+Inf'});

    str = mod_util.format('%s %s %s\n', '# HELP', this.name, this.help);
    str += mod_util.format('%s %s %s\n', '# TYPE', this.name, this.type);

    /*
     * We're digging deep in this loop.
     * The objective of this code is to get each Counter's MetricVec
     * to report its metrics in the prometheus style. But we don't want the
     * Counter itself to do that reporting, since then it would add on the
     * '# TYPE' and '# HELP' headers.
     *
     * In addition, we need to serialize extra fields, like the _sum and _count
     * fields. Both of these fields are reported based on label key/value
     * combinations. That information is stored in each of the Metric objects.
     */
    var promCallback = function (err, metrics) {
        if (err) {
            cb(null, err);
        }
        str += metrics;
    };
    for (counter in this.counters) {
        // Bypass the Counter object's .prometheus function, and instead
        // call it on the metricVector itself.
        this.counters[counter].metricVec.prometheus(promCallback);

        // Get the last object in the metric vector. All of the metrics
        // within a metric vector have the same labels.
        keys = Object.keys(this.counters[counter].metricVec.metrics);
        key = keys[0];
        labelStr = '';
        labelObj = {};
        countLabels = this.counters[counter].metricVec.metrics[key].labels;
        for (label in countLabels) {
            if (label !== 'le') {
                // Ignore the internal 'le' label.
                labelStr += mod_util.format('%s="%s",',
                        label, countLabels[label]);
                labelObj[label] = countLabels[label];
            }
        }
        // Chop off the trailing ','
        labelStr = labelStr.substring(0, labelStr.length - 1);

        // Create the _count metric.
        str += mod_util.format('%s_count{%s}', this.name, labelStr);

        // Append the +Inf value to the _count field, since they are the same.
        str += mod_util.format(' %d\n',
                this.counters[counter].metricVec.metrics[infHash].value);

        // Read the _sum Gauge.
        str += mod_util.format('%s_sum{%s}', this.name, labelStr);
        str += mod_util.format(' %d\n', this.gauge.labels(labelObj).value);
    }

    cb(null, str);
};

/*
 * Returns a Counter with no labels. This is used when an observation is made
 * with no labels.
 */
Histogram.prototype.defaultCounter = function defaultCounter() {
    var opts;
    if (!this.counters['default']) {
        opts = {
            name: this.name,
            help: this.help,
            labels: this.staticLabels
        };
        this.counters['default'] = new lib_counter.Counter(opts);
    }
    return (this.counters['default']);
};

module.exports = {
    Histogram: Histogram
};
